### PR TITLE
feat: Add sideinput generator and watcher, fixes: 1192

### PR DESCRIPTION
**Pull request by chandankumar4**
- Add side input generator script
- Add side input file watcher and publish message using map and reducer

Fixes: https://github.com/numaproj/numaflow/issues/1192

### CODE REVIEW

**Comment by qhuai** [comment]
Should this repo be part of Numaproj? Please also notice [this PR](https://github.com/numaproj/numaflow/pull/1211/files) makes Numaflow depending on images built from this repo. cc @vigith

**Comment by chandankumar4** [comment]
I have pushed the github action changes but didn't see any action running, do we need to enable it ? @vigith

**Comment by KeranYang** [comment]
> I have pushed the github action changes but didn't see any action running, do we need to enable it ?

Could you try removing

```
    tags:
      - v*
```
from the ci workflow file?

File: map-side-input/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,20 @@
+####################################################################################################
```
Comment (KeranYang) [inline map-side-input/Dockerfile:1]: Please add a README.md under `map-side-input` folder to briefly introduce what this UDF does and how to use it. Same for `reduce-side-input`.

File: reduce-side-input/main.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	"strconv"
+
+	sideinputsdk "github.com/numaproj/numaflow-go/pkg/sideinput"
+)
+
+var counter = 0
+
+// handle is the side input handler function.
+func handle(_ context.Context) sideinputsdk.Message {
+	// generate message based on even and odd counter
```
Comment (KeranYang) [inline reduce-side-input/main.go:30]: Do we care if the counter is even or odd for reducer?

**Review by KeranYang — COMMENTED** [review]
nit:

For map, the UDF skips checking input message and only processes side input. reducer skips returning sum if side input data presents.

Although such setup fulfills the goal of testing SI, I am wondering if we could write tests that actually combine map/reduce with SI to process messages, which are more practical.

File: reduce-side-input/main.go (hunk 2 above)
Comment (chandankumar4) [reply reduce-side-input/main.go:30]: no, I'll update the comment

**Comment by chandankumar4** [comment]
> nit:
> 
> For map, the UDF skips checking input message and only processes side input. reducer skips returning sum if side input data presents.
> 
> Although such setup fulfills the goal of testing SI, I am wondering if we could write tests that actually combine map/reduce with SI to process messages, which are more practical.

Done, Thanks for review comment

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        with:
+          registry: quay.io
+          username: ${{ secrets.QUAYIO_USERNAME }}
+          password: ${{ secrets.QUAYIO_PASSWORD }}
+
+      - name: "Step 10: Build docker image"
+        if: github.ref == 'refs/heads/main' || startsWith(github.ref, 'refs/tags/v')
+        run: |
+          docker push "quay.io/numaio/numaproj-contrib/e2e-reduce-sideinput-example:$(basename $GITHUB_REF)"
+          docker push "quay.io/numaio/numaproj-contrib/e2e-reduce-sideinput-udf:$(basename $GITHUB_REF)"
```
Comment (KeranYang) [inline .github/workflows/ci.yaml:126]: Nice!

File: map-side-input/udf/main.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+	"github.com/numaproj/numaflow-go/pkg/mapper"
+	"github.com/numaproj/numaflow-go/pkg/sideinput"
+)
+
+var sideInputName = "myticker"
+var sideInputData []byte
+var mu sync.RWMutex
+
+func mapFn(_ context.Context, _ []string, d mapper.Datum) mapper.Messages {
```
Comment (KeranYang) [inline map-side-input/udf/main.go:35]: The input message is not used. This is not quite practical.

Comment (chandankumar4) [reply map-side-input/udf/main.go:35]: I'm not using key but using value, i.e.: "d mapper.Datum", In line no: 36 and then using "msg" variable in line no:45.

Comment (KeranYang) [reply map-side-input/udf/main.go:35]: Sorry, I missed it.

**Review by KeranYang — APPROVED** [review]

**Pull request merged**